sdk/pinocchio/src/program_error.rs

```diff
@@ -284,6 +284,38 @@ impl ToStr for ProgramError {
     }
 }
 
+/// Require function to check for invariants and error out using custom errors defined in the pinocchio-err crate
+/// Logs the message defined in the msg attribute of the custom error's variants before erroring out
+///
+/// *Arguments* :
+///
+/// `invariant`: The invariant that shouldn't be false
+///
+/// `error`: A variant of the custom error defined in your program, which derives the ErrorCode trait
+///
+/// **`Note`**: **This function cannot be used with the ProgramError enum, as it's only optimized for custom errors for now**
+#[macro_export]
+macro_rules! require {
+    ($invariant:expr, $error:expr $(,)?) => {
+        if !$invariant {
+            // If the error type has a `message()` function, call it.
+            // Otherwise, do nothing.
+            #[allow(unused_variables)]
+            {
+                if false {
+                    // just to scope-check type
+                } else {
+                    #[allow(unused_unsafe)]
+                    unsafe {
+                        $crate::msg!($error.message());
+                    }
+                }
+            }
```

Comment:
This is super strange. I am not sure why it is not written just like:
```rust
$crate::msg!($error.message());
```
Also, it is inefficient in terms of program size since it will generate a call to the log syscall in every use of `require!`. It would be much better to follow a pattern where any error is printed at the end of the `process_instruction`, similar to p-token: https://github.com/solana-program/token/blob/main/pinocchio/program/src/entrypoint.rs#L258